### PR TITLE
Random matching

### PR DESCRIPTION
**Pull request by kurtcarpenter**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -12,6 +12,8 @@ var config = require('./config')
 var publicRoutes = require('./routes/public')
 var classRoutes = require('./routes/class')
 
+var matcher = require('./lib/matcher/matcher')
```
Comment (kurtcarpenter) [inline app.js:15]: for ensuring it parses okay

File: lib/matcher/strategies/random.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  *   teams: [team1: {}, team2: {}...],
+  *   details: 'Two teams were overfilled to match all students.'
+  * }
+*/
+
+var logger = require('winston')
+
+exports.formTeams = function (roster, teams, min, max) {
+  if (!roster || !teams || !Array.isArray(roster) || !Array.isArray(teams)) {
+    logger.warn('Invalid input to random.formTeams')
```
Comment (supersam654) [inline lib/matcher/strategies/random.js:18]: Return

File: lib/matcher/strategies/random.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  }
+  if (!min || !max) {
+    logger.warn('Invalid size parameters to random.formTeams')
```
Comment (supersam654) [inline lib/matcher/strategies/random.js:21]: Return

File: lib/matcher/matcher.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    cb('Invalid matching strategy: ' + strategy)
+  }
+  if (!classObj.roster || !classObj.minGroup || !classObj.maxGroup) {
+    cb('Malformed class object')
+  }
+
+  switch (strategy) {
+    case 'RANDOM':
+      var result = StrategyRandom.formTeams(classObj.roster, preserveTeams ? classObj.teams : [])
+      break
```
Comment (supersam654) [inline lib/matcher/matcher.js:33]: Please add a default case with a `logger.warn` here.

File: lib/matcher/strategies/random.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  // initialize the teams array with n buckets
+  for (var i = 0; i < totalTeams - finalTeams.length; i++) {
+    finalTeams.push({
+      members: []
+    })
+  }
+
+  // put people into buckets
+  for (var i = 0; i < roster.length; i++) {
+    var hash = hashStudent(roster[i]) % totalTeams
```
Comment (supersam654) [inline lib/matcher/strategies/random.js:39]: Please use shuffle instead: http://stackoverflow.com/a/2450976/868533

File: lib/matcher/matcher.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  }
+})
+
+exports.formTeams = function (classObj, strategy, preserveTeams, cb) {
+  if (!cb || typeof cb !== 'function') {
+    logger.warn('Failed to pass callback to matcher.formTeams')
+    return
+  }
+  if (!exports.strategies[strategy]) {
+    cb('Invalid matching strategy: ' + strategy)
```
Comment (supersam654) [inline lib/matcher/matcher.js:24]: `return cb`

File: lib/matcher/matcher.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  }
+  if (!classObj.roster || !classObj.minGroup || !classObj.maxGroup) {
+    cb('Malformed class object')
```
Comment (supersam654) [inline lib/matcher/matcher.js:27]: `return cb`

**Review by supersam654 — CHANGES_REQUESTED** [review]

**Pull request merged**